src/Microsoft.TestPlatform.CrossPlatEngine/EventHandlers/PathConverter.cs

```diff
@@ -0,0 +1,145 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.CommunicationUtilities;
+using System.Collections.Generic;
+using System.Linq;
+using ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel.Client;
+using System.Collections.ObjectModel;
+using Microsoft.VisualStudio.TestPlatform.Utilities.Helpers.Interfaces;
+using System.IO;
+
+internal class PathConverter : IPathConverter
+{
+    // The path on this computer to which we deployed the test dll and test runner
+    private readonly string _deploymentPath = "";
+    // The path on the remote system where test dll was originally placed, and from which we
+    // copied it to this system. For vstest.console, which is on the other side of this, the names
+    // are inverted, it sends us their local path, and thinks about our local path as remote.
+    private readonly string _originalPath = "";
+
+    public PathConverter(string originalPath, string deploymentPath, IFileHelper fileHelper)
+    {
+        _originalPath = fileHelper.GetFullPath(originalPath).TrimEnd('\\').TrimEnd('/') + Path.DirectorySeparatorChar;
+        _deploymentPath = fileHelper.GetFullPath(deploymentPath).TrimEnd('\\').TrimEnd('/') + Path.DirectorySeparatorChar;
+    }
+
+    public string UpdatePath(string path, PathConversionDirection updateDirection)
+    {
+        string find;
+        string replaceWith;
+        if (updateDirection == PathConversionDirection.Receive)
+        {
+            // Request is incoming, the path that is local to the sender (for us that is "remote" path)
+            // needs to be replaced with our path
+            find = _originalPath;
+            replaceWith = _deploymentPath;
+        }
+        else
+        {
+            find = _deploymentPath;
+            replaceWith = _originalPath;
+        }
+
+        var result = path?.Replace(find, replaceWith);
+        return result;
+    }
+
+    public IEnumerable<string> UpdatePaths(IEnumerable<string> enumerable, PathConversionDirection updateDirection)
+    {
+        var updatedPaths = enumerable.Select(i => UpdatePath(i, updateDirection)).ToList();
```

Comment:
I would love to do that this way, but the object contract is public, and the only thing I can do without changing it, is making this work in a terrible side-effectful way. 

The select here will work just fine, ToList forces enumeration, but I changed it to use tolist and foreach in all places to make it more obvious that we do something that has side effects.

We don't cast down to anything here, I don't see where this would fail. Care to give an example?

Reply:
That is "casting" up imho, IEnumerable is higher in the inheritance tree than List. But yeah, what you show is stupid design that is trying to be overly generic, and uses IEnumerable without using the lazy nature of it. I am not trying to do that here, I just use the types that are incoming in the request so I can go from `var serialized = Serializer.Serialize(dto)` to `var serialized = Serializer.Serialize(_pathConverter.Update(dto))`.

As for the side-effects, I am updating dtos that have a non-default constructor and non-settable properties, so I can't just create a new one with the updated data. I would love that, but instead I am forced to use side-effects to change the values in the incoming data (in most cases). Without that I would have to change the public api of those dtos, so they allow me to recreate them.